d2/src/main/java/com/linkedin/d2/balancer/subsetting/SubsettingState.java

```diff
@@ -0,0 +1,164 @@
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.d2.balancer.subsetting;
+
+import com.linkedin.d2.balancer.clients.TrackerClient;
+import java.net.URI;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.http.annotation.GuardedBy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * State of cluster subsetting
+ */
+public class SubsettingState
+{
+  private static final Logger LOG = LoggerFactory.getLogger(SubsettingState.class);
+  private final Object _lock = new Object();
+  private final SubsettingStrategyFactory _subsettingStrategyFactory;
+
+  /**
+   * Map from serviceName => Partition Id => subset map
+   */
+  @GuardedBy("_lock")
+  private final Map<String, Map<Integer, Map<URI, TrackerClient>>> _weightedSubsetsCache;
+
+  /**
+   * Map from serviceName => Partition Id => all potential clients
+   */
+  @GuardedBy("_lock")
+  private final Map<String, Map<Integer, Set<URI>>> _potentialClientsMap;
+
+  @GuardedBy("_lock")
+  private final Map<String, Integer> _minClusterSubsetSizeMap;
+
+  @GuardedBy("_lock")
+  private long _version;
+
+  @GuardedBy("_lock")
+  private long _peerClusterVersion;
+
+  public SubsettingState(SubsettingStrategyFactory subsettingStrategyFactory)
+  {
+    _version = -1;
+    _peerClusterVersion = -1;
+    _subsettingStrategyFactory = subsettingStrategyFactory;
+    _weightedSubsetsCache = new HashMap<>();
+    _potentialClientsMap = new HashMap<>();
+    _minClusterSubsetSizeMap = new HashMap<>();
+  }
+
+  public long getPeerClusterVersion(String serviceName,
+      int minClusterSubsetSize,
+      int partitionId)
+  {
+    SubsettingStrategy<URI> subsettingStrategy = _subsettingStrategyFactory.get(serviceName, minClusterSubsetSize, partitionId);
+
+    if (subsettingStrategy == null)
+    {
+      return -1;
+    }
+
+    return subsettingStrategy.getPeerClusterVersion();
+  }
+
+  public Map<URI, TrackerClient> getClientsSubset(String serviceName,
+      int minClusterSubsetSize,
+      int partitionId,
+      Map<URI, TrackerClient> potentialClients,
+      long version)
+  {
+    SubsettingStrategy<URI> subsettingStrategy = _subsettingStrategyFactory.get(serviceName, minClusterSubsetSize, partitionId);
+
+    if (subsettingStrategy == null)
+    {
+      return potentialClients;
+    }
+
+    long peerClusterVersion = subsettingStrategy.getPeerClusterVersion();
+
+    synchronized (_lock)
+    {
+      if (isCacheValid(version, peerClusterVersion, minClusterSubsetSize, serviceName))
+      {
+        Map<Integer, Map<URI, TrackerClient>> serviceCache = _weightedSubsetsCache.get(serviceName);
+        if (serviceCache != null && serviceCache.containsKey(partitionId))
+        {
+          return serviceCache.get(partitionId);
+        }
+      }
```

Comment:
If we have multiple downstream services, now we have a new request calling service A. When we get the subset hosts for service A, we found that the peerClusterVersion is changed, so we updated the cache for service A.

Now there is a new request to service B, when we process the request to service B, minClusterSubsetSize remains the same, and the cached peerClusterVersion is up to date since we updated during service A request. In that case, will we not update the subset for service B?

Reply:
The question is inspired by your reply to the other question. I was thinking about if the cache update should be done by request thread or uri listener thread. When I said why don't we push the update, I was thinking about if we should update as long as we receive an update from zk, but I found it's not that easy.

Reply:
Fixed in the latest commit